reducer/src/main/java/com/graphicsfuzz/reducer/reductionopportunities/LiteralToUniformReductionOpportunity.java

```diff
@@ -16,27 +16,108 @@
-  private final PipelineInfo pipelineInfo;
+  private final ConstantExpr literalExpr;
+  private final TranslationUnit translationUnit;
+  private final ShaderJob shaderJob;
 
-  LiteralToUniformReductionOpportunity(Expr literalExpr,
-                                       PipelineInfo pipelineInfo,
+  LiteralToUniformReductionOpportunity(ConstantExpr literalExpr,
+                                       TranslationUnit translationUnit,
+                                       ShaderJob shaderJob,
                                        VisitationDepth depth) {
+
     super(depth);
+
     this.literalExpr = literalExpr;
-    this.pipelineInfo = pipelineInfo;
+    this.translationUnit = translationUnit;
+    this.shaderJob = shaderJob;
   }
 
   @Override
   void applyReductionImpl() {
-    // TODO: ...
+
+    final Number numericValue;
+    final String arrayName;
+    final BasicType basicType;
+
+    // If the uniform array doesn't exist in the pipeline info, adds an empty array.
+    if (literalExpr instanceof IntConstantExpr) {
+      arrayName = Constants.INT_LITERAL_UNIFORM_VALUES;
+      basicType = BasicType.INT;
+      numericValue = ((IntConstantExpr) literalExpr).getNumericValue();
+    } else {
+      arrayName = Constants.FLOAT_LITERAL_UNIFORM_VALUES;
+      basicType = BasicType.FLOAT;
+      numericValue = Float.valueOf(((FloatConstantExpr) literalExpr).getValue());
+    }
+
+    if (!shaderJob.getPipelineInfo().hasUniform(arrayName)) {
+      shaderJob.getPipelineInfo().addUniform(arrayName, basicType,
+          Optional.of(0), new ArrayList<>());
+      shaderJob.getPipelineInfo().addUniformBinding(arrayName, false,
+          shaderJob.getPipelineInfo().getUnusedBindingNumber());
+    }
+
+    final int index;
+    final List<Number> values = this.shaderJob.getPipelineInfo().getArgs(arrayName);
+
+    if (values.contains(numericValue)) {
+      index = values.indexOf(numericValue);
+    } else {
+      index = shaderJob.getPipelineInfo().appendValueToUniform(arrayName, numericValue);
+    }
+
+    // Adds the new array declaration to the current translation unit and updates the size
+    // in every translation unit where the declaration is present.
+    final int arraySize = shaderJob.getPipelineInfo().getArgs(arrayName).size();
+    final ArrayInfo arrayInfo = new ArrayInfo(new IntConstantExpr(String.valueOf(arraySize)));
+    final VariableDeclInfo variableDeclInfo = new VariableDeclInfo(arrayName,
+        arrayInfo, null);
+    final VariablesDeclaration arrayDecl = new VariablesDeclaration(
+        new QualifiedType(basicType, Arrays.asList(TypeQualifier.UNIFORM)), variableDeclInfo
+    );
+
+    if (!this.translationUnit.hasUniformDeclaration(arrayName)) {
+      this.translationUnit.addDeclaration(arrayDecl);
+    }
+
+    for (TranslationUnit tu: shaderJob.getShaders()) {
+      if (tu.hasUniformDeclaration(arrayName)) {
+        tu.updateTopLevelDeclaration(arrayDecl, tu.getUniformDeclaration(arrayName));
```

Comment:
Is the existing translation unit guaranteed to have the array uniform?

Example: suppose our shader job has a vertex shader and a fragment shader, and neither originally have the array uniform.

We then replace a literal in the fragment shader.

As a result, we will add the array uniform to the fragment shader (via the code above).  Then we will go through all the shaders and do this replacement - but there will be nothing to replace in the vertex shader.

Can you make sure we have a test for that case?

Reply:
If I am correct then I think you should replace the conditional that adds the array uniform if not already present and the loop that does replacement with a single loop that considers each translation unit in turn, and adds the array if not present or replaces it if present.

Reply:
I am not so sure Ally. I think we should only ever _add_ the declaration to the target `tu`, if needed. We should loop through _all_ other `tu`s to make sure we update their size, if needed.

Reply:
The current code _adds_ to the target `tu`. It updates every other `tu` but only if the declaration is present. It looks good to me.

Reply:
Oh yes - you're absolutely right.  @amakin-siru - please ignore my suggestion.